src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableSortedSet`1.cs

```diff
@@ -1102,10 +1102,39 @@ private ImmutableSortedSet<T> LeafToRootRefill(IEnumerable<T> addedItems)
             // and can index into that sequence like a list, so the limited
             // garbage produced is a temporary mutable data structure we use
             // as a reference when creating the immutable one.
-            // The (mutable) SortedSet<T> is much faster at constructing its collection 
-            // when passed a sequence into its constructor than into its Union method.
-            var sortedSet = new SortedSet<T>(this.Concat(addedItems), this.KeyComparer);
-            Node root = Node.NodeTreeFromSortedSet(sortedSet);
+
+            // Produce the initial list containing all elements, including any duplicates.
+            List<T> list;
+            if (this.IsEmpty)
+            {
+                list = new List<T>(addedItems);
+                if (list.Count == 0)
+                {
+                    return this;
+                }
+            }
+            else
+            {
+                list = new List<T>(this);
+                list.AddRange(addedItems);
+            }
+            Debug.Assert(list.Count > 0);
+
+            // Sort the list and remove duplicate entries.
+            IComparer<T> comparer = this.KeyComparer;
+            list.Sort(comparer);
```

Comment:
:exclamation: The combination of an unstable sort here with the use of a list of all elements when the current set is not empty has what I believe is a particularly bad side effect. Unlike methods using `ISet.UnionWith`, the behavior of `ImmutableSortedSet` with respect to duplicate entries is *unspecified*. The `ISet.UnionWith` method explicitly states that entries already in the collection take precedence in the face of duplicates to a `Union` operation.